core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java

```diff
@@ -388,6 +390,30 @@ private void setRoleName(ApplicationInsightsXmlConfiguration userConfiguration,
         }
     }
 
+    private void setRoleInstance(ApplicationInsightsXmlConfiguration userConfiguration,
+                             TelemetryConfiguration configuration) {
+        try {
+            String roleInstance;
+
+            // try to find the role instance in ApplicationInsights.xml
+            if (userConfiguration != null) {
+                roleInstance = userConfiguration.getRoleInstance();
+                if (roleInstance == null) {
+                    return;
+                }
+
+                roleInstance = roleInstance.trim();
+                if (roleInstance.length() == 0) {
+                    return;
+                }
+
+                configuration.setRoleInstance(roleInstance);
+            }
+        } catch (Exception e) {
+            InternalLogger.INSTANCE.error("Failed to set role instance: '%s'", e.toString());
+        }
+    }
```

Comment:
Are the environment variables a "codeless agent only" mechanism? If not, perhaps they should be read in this class instead/also.